flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java

```diff
@@ -539,20 +544,21 @@ private void requestYarnContainer() {
 			RM_REQUEST_PRIORITY);
 	}
 
-	private ContainerLaunchContext createTaskExecutorLaunchContext(Resource resource, String containerId, String host)
-			throws Exception {
+	private ContainerLaunchContext createTaskExecutorLaunchContext(
+		TaskExecutorResourceSpec tmResourceSpec,
```

Comment:
btw, why do we need to pass `tmResourceSpec` if there is already a class field `taskExecutorResourceSpec`?
also, looks like `taskExecutorResourceSpec` is not adjusted with the really created container memory, previously taken from `resource.getMemory()`. Maybe we do not need to pre-create `taskExecutorResourceSpec`.

Reply:
True, we should use the class field directly.
We cannot adjust `taskExecutorResourceSpec` to `resource.getMemory()`. Because we generate profiles for pending slots before containers are allocated, based on the memory sizes derived from configuration. If we then change the `taskExecutorResourceSpec` to `resource.getMemory()`, it may generated different JVM parameters and dynamic configurations, of which the latter will be used for generating slot profiles on the TM side. As a result, we may have different slot profiles for pending slots on RM side and actual slots on TM side.

Reply:
ok, makes sense. then the unused `resource` function parameter should be removed and/or it would be nice to have a check that allocated container has at least the requested memory as it is basically our assumption.

Reply:
In future, we probably can take this into account with `PendingTaskExecutorResource` in FLIP-56 where we could try to match it with the registration by some id.

Reply:
True, I'll remove the unused `resource`.

For the memory size checking, it is guaranteed by Yarn that the allocated container should have at least requested memory size, or otherwise the request will not be satisfied. But I think it doesn't harm to add an assertion for double checking.

Agree that in future we can have some smarter way to match pending task executor resource, so we can leverage the extra memory allocated in the container.